### PR TITLE
[lldb] Make logging machinery type-safe

### PR DESCRIPTION
**Pull request by medismailben**
This patch makes use of c++ type checking and scoped enums to make
logging statements shorter and harder to misuse.

Defines like LIBLLDB_LOG_PROCESS are replaces with LLDBLog::Process.
Because it now carries type information we do not need to worry about
matching a specific enum value with the right getter function -- the
compiler will now do that for us.

The main entry point for the logging machinery becomes the GetLog
(template) function, which will obtain the correct Log object based on
the enum type. It achieves this through another template function
(LogChannelFor<T>), which must be specialized for each type, and should
return the appropriate channel object.

This patch also removes the ability to log a message if multiple
categories are enabled simultaneously as it was unused and confusing.

This patch does not actually remove any of the existing interfaces. The
defines and log retrieval functions are left around as wrappers around
the new interfaces. They will be removed in follow-up patch.

Differential Revision: https://reviews.llvm.org/D117490

### CODE REVIEW

**Comment by medismailben** [comment]
This depends on #3958

**Comment by medismailben** [comment]
@swift-ci test

**Review by kastiglione — APPROVED** [review]
thanks for picking this!

File: lldb/source/Plugins/Language/Swift/LogChannelSwift.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+static std::string g_swift_log_buffer;
+
+Log::Channel LogChannelSwift::g_channel(g_categories, SwiftLog::Health);
+
+template <> Log::Channel &lldb_private::LogChannelFor<SwiftLog>() {
+  return g_channel;
+}
+
+void LogChannelSwift::Initialize() {
```
Comment (kastiglione) [inline lldb/source/Plugins/Language/Swift/LogChannelSwift.cpp:29]: I don't see where this `Initialize()` is called.

**Comment by medismailben** [comment]
@swift-ci test

File: lldb/source/Plugins/Language/Swift/LogChannelSwift.cpp (hunk 1 above)
Comment (medismailben) [reply lldb/source/Plugins/Language/Swift/LogChannelSwift.cpp:29]: Resolved in 0f528214795aa15b4cc250ba41ce271ca6af2e70

**Comment by medismailben** [comment]
@swift-ci test

**Comment by JDevlieghere** [comment]
Thanks Ismail!

**Review by JDevlieghere — APPROVED** [review]

**Comment by medismailben** [comment]
@swift-ci test

**Comment by medismailben** [comment]
@swift-ci test linux

**Comment by medismailben** [comment]
#3906 introduces a new merge conflict 🥲 @swift-ci test again

**Comment by medismailben** [comment]
@swift-ci test and merge

**Comment by medismailben** [comment]
@swift-ci test

**Comment by medismailben** [comment]
@swift-ci test

**Comment by compnerd** [comment]
This caused a regression in the build; you added a new source file but didn't correctly update the build system to indicate that it is an optional source file, which breaks the LLVM build.

**Comment by medismailben** [comment]
Hi @compnerd, I'm just seeing this. Sorry I missed that and thanks for fixing it!

**Pull request merged**